Dspace 7 shibboleth (Angular)

atarix83 opened the pull request:
This PR is based on shibboleth implementation made by @Flusspferd123 in this [429](https://github.com/DSpace/dspace-angular/pull/429) and provide changes discussed in [DS-4396](https://jira.lyrasis.org/browse/DS-4396)

It needs to be run over [this rest implemenation](https://github.com/DSpace/DSpace/pull/2651)

this PR has been already merged in the dspace-cris 7 codebase so it is possible to see the behavior here: [https://dspacecris7.4science.cloud/home](https://dspacecris7.4science.cloud/home)

atarix83 commented:
@benbosman Did you test this branch over the rest implemented in [this PR](https://github.com/DSpace/DSpace/pull/2651) ?
Because in REST PR 'www-authenticate' is sent also in response to a GET to /server/api/authn/status, and is how the header is retrieved in this angular PR

benbosman commented:
@atarix83 it was probably an error in my build of https://github.com/DSpace/DSpace/pull/2651 or a caching issue. It only worked from https://dspace7-ben.atmire.com/server/login.html, but I've restarted REST and now it displays the login options in Angular. I will continue my testing

paulo-graca commented on src/app/core/auth/auth.actions.ts:
```diff
@@ -308,6 +325,40 @@ export class ResetAuthenticationMessagesAction implements Action {
+  }
+}
+
+export class RetrieveAuthMethodsErrorAction implements Action {
```
any particular reason for excluding Typescript docs here?

atarix83 replied:
added

atarix83 commented on src/app/core/auth/auth.actions.ts:
```diff
@@ -137,11 +141,11 @@ export class CheckAuthenticationTokenAction implements Action {
  */
-export class CheckAuthenticationTokenErrorAction implements Action {
-  public type: string = AuthActionTypes.CHECK_AUTHENTICATION_TOKEN_ERROR;
+export class CheckAuthenticationTokenCookieAction implements Action {
```
CheckAuthenticationTokenErrorAction doesn't exist anymore, please see how it's refatored [here](https://github.com/DSpace/dspace-angular/pull/568/files#diff-07178cc516a358e3cc31b3d3cb614685R85)

atarix83 commented:
@benbosman 

> I have tested this functionality with Angular and REST on the same server. This worked, but I noticed it didn't always return to the page I start from. I started the authentication from a collection page, and was redirected to the homepage. The redirectUrl parameter was incorrect.
I did originally open the homepage, click through to the collection page, and started authentication next

I fixed the redirect url and replaced with the current page url

> I noticed that using 2 separate servers, you'll need Apache settings similar to DSpace/DSpace#2651 (but they didn't work for http://localhost:3000)

You are right to make it work also the port should be included in the configuration, I'll change it in the rest PR

> the authentication-shibboleth.lazysession.loginurl needs to be a full URL. If the location from authentication-shibboleth.lazysession.loginurl is a relative URL, Angular should add the REST host.

I think this should be resolved on rest side. it should provide the correct url for shibboleth authentication. I'll change it in the rest PR

benbosman commented:
@atarix83 thanks for the updates.

I can confirm I am now being redirected to my last page correctly after authenticating with Shibboleth

I can also confirm I was able to test with a local Angular on localhost:3000 using the settings from
https://github.com/DSpace/DSpace/pull/2651#issuecomment-584576733

I agree authentication-shibboleth.lazysession.loginurl should be solved on REST, and included my input on https://github.com/DSpace/DSpace/pull/2651#discussion_r381200372

I noticed an empty git file is added, I assume this was unintentional?

I will check the code changes as well, haven't started that part of the review, but I noticed @paulo-graca is already reviewing the code

benbosman commented:
One more thing I noticed while testing this functionality.
When testing using an Angular UI with a different hostname, I receive:
> A cookie associated with a cross-site resource at http://dspace7-shibtest.atmire.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032.

Not sure whether that's related

atarix83 commented:
@benbosman @paulo-graca 
>I've tested this branch using DSpace/DSpace#2651 in the backend and with Shibboleth enabled. I used a local Angular (with the CORS config).
When opening the "log-in" overlay, no authentication methods are displayed (neither password nor shibboleth login) while I had both enabled

This issue has been solved now

paulo-graca commented:
This PR has a code conflict. Can you please fix them so I can redo my tests?

atarix83 commented:
@paulo-graca rebase done

tdonohue commented:
Merging as this is at +2 and a good enough initial implementation. If any further updates or bug fixes are needed, those can be done in a follow-up PR.

As noted above, the instructions for Shib setup have been drafted at https://wiki.lyrasis.org/display/DSPACE/DSpace+7+Shibboleth+Configuration
